File: src/snapdragon_npu_audio/audio_frame.py
from __future__ import annotations

from dataclasses import dataclass
from collections.abc import Iterable


Sample = float
StereoSample = tuple[Sample, Sample]


@dataclass(frozen=True)
class AudioFrame:
    """A short interleaved stereo block normalized to -1.0..1.0 floats."""

    sample_rate: int
    samples: tuple[StereoSample, ...]

    def __post_init__(self) -> None:
        if self.sample_rate <= 0:
            raise ValueError("sample_rate must be positive")

    @property
    def duration_seconds(self) -> float:
        return len(self.samples) / self.sample_rate

    @property
    def peak(self) -> float:
        return max((max(abs(left), abs(right)) for left, right in self.samples), default=0.0)

    @property
    def rms(self) -> float:
        if not self.samples:
            return 0.0
        energy = sum(left * left + right * right for left, right in self.samples)
        return (energy / (len(self.samples) * 2)) ** 0.5

    def map_samples(self, gain_left: float, gain_right: float) -> "AudioFrame":
        return AudioFrame(
            sample_rate=self.sample_rate,
            samples=tuple((left * gain_left, right * gain_right) for left, right in self.samples),
        )

    def apply_gain_db(self, gain_db: float) -> "AudioFrame":
        gain = 10.0 ** (gain_db / 20.0)
        return self.with_samples((left * gain, right * gain) for left, right in self.samples)

    def with_samples(self, samples: Iterable[StereoSample]) -> "AudioFrame":
        return AudioFrame(
            sample_rate=self.sample_rate,
            samples=tuple((clamp_sample(left), clamp_sample(right)) for left, right in samples),
        )

    def mono(self) -> list[float]:
        return [(left + right) * 0.5 for left, right in self.samples]

    @classmethod
    def from_mono(cls, sample_rate: int, samples: Iterable[Sample]) -> "AudioFrame":
        return cls(sample_rate=sample_rate, samples=tuple((sample, sample) for sample in samples))


def clamp_sample(sample: float, limit: float = 1.0) -> float:
    return max(-limit, min(limit, sample))

File: src/dsp/audio_frame.cpp
#include "dsp/audio_frame.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace snae::dsp {

AudioFrame::AudioFrame(int sample_rate, std::vector<StereoSample> samples)
    : sample_rate_(sample_rate), samples_(std::move(samples)) {
    if (sample_rate_ <= 0) {
        throw std::invalid_argument("sample rate must be positive");
    }
}

AudioFrame::AudioFrame(std::size_t samples_per_channel, int sample_rate)
    : AudioFrame(sample_rate, std::vector<StereoSample>(samples_per_channel)) {}

int AudioFrame::sample_rate() const noexcept {
    return sample_rate_;
}

std::size_t AudioFrame::size() const noexcept {
    return samples_.size();
}

bool AudioFrame::empty() const noexcept {
    return samples_.empty();
}

const std::vector<StereoSample>& AudioFrame::samples() const noexcept {
    return samples_;
}

std::vector<StereoSample>& AudioFrame::samples() noexcept {
    return samples_;
}

void AudioFrame::apply_gain(float gain) {
    for (auto& sample : samples_) {
        sample.left *= gain;
        sample.right *= gain;
    }
}

float measurePeak(const AudioFrame& frame) {
    float peak = 0.0F;
    for (const auto& sample : frame.samples()) {
        peak = std::max({peak, std::abs(sample.left), std::abs(sample.right)});
    }
    return peak;
}

float measureRms(const AudioFrame& frame) {
    if (frame.empty()) {
        return 0.0F;
    }

    double sum_squares = 0.0;
    for (const auto& sample : frame.samples()) {
        sum_squares += static_cast<double>(sample.left) * sample.left;
        sum_squares += static_cast<double>(sample.right) * sample.right;
    }
    return static_cast<float>(std::sqrt(sum_squares / static_cast<double>(frame.size() * 2U)));
}

}  // namespace snae::dsp
